File: ios/InvitationLink.swift
import Foundation

/// Parses canonical patient-invitation Universal Links:
/// `https://cbtipul.com/invite/<TOKEN>`.
///
/// Does not claim the invitation or talk to the backend.
enum InvitationLink {
    static let host = "cbtipul.com"
    static let pathPrefix = "/invite/"

    static func token(from url: URL) -> String? {
        token(scheme: url.scheme, host: url.host(), path: url.path)
    }

    static func token(scheme: String?, host: String?, path: String?) -> String? {
        guard scheme?.lowercased() == "https" else { return nil }
        guard host?.lowercased() == Self.host else { return nil }
        let path = path ?? ""
        guard path.hasPrefix(pathPrefix) else { return nil }
        let token = String(path.dropFirst(pathPrefix.count))
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        if token.isEmpty || token.contains("/") { return nil }
        return token
    }
}

File: android/app/src/main/java/com/cbtipul/app/invite/InvitationLink.kt
package com.cbtipul.app.invite

import android.net.Uri

/**
 * Parses canonical patient-invitation App Links:
 * `https://cbtipul.com/invite/<TOKEN>`.
 *
 * Does not claim the invitation or talk to the backend.
 */
object InvitationLink {
    const val HOST = "cbtipul.com"
    const val PATH_PREFIX = "/invite/"

    fun tokenFrom(uri: Uri?): String? {
        if (uri == null) return null
        return tokenFrom(uri.scheme, uri.host, uri.path)
    }

    fun tokenFrom(scheme: String?, host: String?, path: String?): String? {
        if (scheme != "https") return null
        if (host != HOST) return null
        val normalizedPath = path.orEmpty()
        if (!normalizedPath.startsWith(PATH_PREFIX)) return null
        val token = normalizedPath.removePrefix(PATH_PREFIX).trimEnd('/')
        if (token.isEmpty() || token.contains('/')) return null
        return token
    }
}
